### ISSUE TITLE
HAProxy Reload - Old Processes not finishing

### ISSUE THREAD

**Issue by mtraceable** [haproxy/haproxy#1876]
### Detailed Description of the Problem

When using SPOE the the old processes are not exiting on Reload. They keep staying indefinitely with their sockets being in the CLOSED_WAIT state ecample output:

ubuntu@ip-37-0-2-146:~$ ps -eaf|grep -i hap
root        4818       1  0 21:29 ?        00:00:00 /usr/sbin/haproxy -sf 4820 4821 4822 4823 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     4820    4818  0 21:29 ?        00:00:02 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     4821    4818  0 21:29 ?        00:00:04 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     4822    4818  0 21:29 ?        00:00:18 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     4823    4818  0 21:29 ?        00:00:14 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     6945    4818  0 22:23 ?        00:00:00 /usr/sbin/haproxy -sf 4820 4821 4822 4823 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     6946    4818  1 22:23 ?        00:00:00 /usr/sbin/haproxy -sf 4820 4821 4822 4823 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     6947    4818  8 22:23 ?        00:00:01 /usr/sbin/haproxy -sf 4820 4821 4822 4823 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     6948    4818 20 22:23 ?        00:00:03 /usr/sbin/haproxy -sf 4820 4821 4822 4823 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock

The sockets show up as:
ubuntu@ip-37-0-2-146:~$ sudo netstat -anpto | grep "4820 \| 4821 \| 4822 \| 4823"
tcp        1      0 37.0.2.146:53188        37.0.2.55:8080          CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:49220        37.0.2.173:8080         CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:80           37.0.2.19:47894         CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:80           37.0.2.152:42330        CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:80           37.0.2.152:41638        CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        0      0 37.0.2.146:49656        37.0.2.173:8080         CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:49230        37.0.2.173:8080         CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:80           37.0.2.19:47742         CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:80           37.0.2.152:41350        CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:80           37.0.1.239:49136        CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:49434        37.0.2.173:8080         CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:49022        37.0.2.173:8080         CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:53062        37.0.2.55:8080          CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:80           37.0.2.152:51310        CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        0      0 37.0.2.146:53282        37.0.2.55:8080          CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
tcp        1      0 37.0.2.146:80           37.0.2.152:42768        CLOSE_WAIT  4823/haproxy         off (0.00/0/0)
.....()

### Expected Behavior

Old processes, eg:
haproxy     4820    4818  0 21:29 ?        00:00:02 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     4821    4818  0 21:29 ?        00:00:04 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     4822    4818  0 21:29 ?        00:00:18 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy     4823    4818  0 21:29 ?        00:00:14 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock

should finish their processing and exit.

### Steps to Reproduce the Behavior

- Using the attached configurations start the HA Proxy
- Send it some good amount of load. In my case around 800 requests per second
- Perform Reload

### What is your configuration?

```haproxy
haproxy.cfg:
TSPOEglobal
                log /dev/log    local0
                log /dev/log    local1 notice
                chroot /var/lib/haproxy
                stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
                stats timeout 30s
                user haproxy
                group haproxy
                daemon
                tune.bufsize 131072
                maxconn 1000
                nbproc 4
                cpu-map  1 0
                cpu-map  2 1
                cpu-map  3 2
                cpu-map  4 3
                tune.ssl.cachesize 40000
                tune.ssl.default-dh-param 2048


                # Default SSL material locations
                ca-base /etc/ssl/certs
                crt-base /etc/ssl/private

                # See: https://ssl-config.mozilla.org/#server=haproxy&server-version=2.0.3&config=intermediate
                ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
                ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
                ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets

frontend stats
                bind *:9080
                maxconn     1000
                http-request disable-l7-retry unless METH_GET
                mode http
                stats uri /stats
                stats show-legends
                # ********** TSPOE CONFIG ***********
                filter spoe engine tspoe config /usr/local/etc/tspoe/tspoe.conf

                # instruct haproxy to make an RPC call to the tspoe agent on request
                # and responses
                http-request send-spoe-group tspoe tspoeai-reqcap-grp
                http-response send-spoe-group tspoe tspoeai-rescap-grp

                # deny requests that the tspoe-agent tells us to block
                http-request deny deny_status 403 if { var(sess.tspoe.block_request) -m bool }
                # *********** END OF TSPOE CONFIG *************
                no log

frontend localhost
                bind *:80
                maxconn     1000
                http-request disable-l7-retry unless METH_GET
                bind *:443 ssl crt /etc/haproxy/certs/chained.pem ssl-min-ver TLSv1.1 ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256 prefer-client-ciphers
                # ********** TSPOE CONFIG ***********
                # Load tspoe's config file.
                filter spoe engine tspoe config /usr/local/etc/tspoe/tspoe.conf

                # instruct haproxy to make an RPC call to the tspoe agent on request
                # and responses
                http-request send-spoe-group tspoe tspoeai-reqcap-grp
                http-response send-spoe-group tspoe tspoeai-rescap-grp

                # deny requests that the tspoe-agent tells us to block
                http-request deny deny_status 403 if { var(sess.tspoe.block_request) -m bool }
                # *********** END OF TSPOE CONFIG *************
                mode http
                default_backend nodes

backend nodes
                mode http
                balance roundrobin
                option forwardfor

                # ********** TSPOE CONFIG ***********
                http-request set-header traceparent %[var(sess.tspoe.trace_context.traceparent)]
                # *********** END OF TSPOE CONFIG *************

                server javaechoserver1 37.0.2.173:8080 check
                server javaechoserver2 37.0.2.55:8080 check

                http-request set-header X-Forwarded-Port %[dst_port]
                http-request add-header X-Forwarded-Proto https if { ssl_fc }


backend tspoe
               mode tcp
               balance roundrobin
               timeout connect 5s  # greater than hello timeout
               timeout server  3m  # greater than idle timeout
               option spop-check

                # set up connectivity to the tspoe agent
               server agent localhost:5444 check


defaults
                log     global
                mode    http
                option  httplog
                option  dontlognull
                errorfile 400 /etc/haproxy/errors/400.http
                errorfile 403 /etc/haproxy/errors/403.http
                errorfile 408 /etc/haproxy/errors/408.http
                errorfile 500 /etc/haproxy/errors/500.http
                errorfile 502 /etc/haproxy/errors/502.http
                errorfile 503 /etc/haproxy/errors/503.http
                errorfile 504 /etc/haproxy/errors/504.http

                # It seems with current configuration we don't need it https://cbonte.github.io/haproxy-dconv/configuration-1.5.html#4-option%20persist
                # option persist
                option redispatch
                option forwardfor

                option http-keep-alive
                #pretend keep-alive closes connection on backend, but we want to keep it open for keep-alive interval
                #option http-pretend-keepalive
                timeout http-keep-alive 30s

                # Connection timeout is to accomodate losses in TCP packets, it should be a few ms in general in LAN
                timeout connect 5s
                timeout check   5s
                timeout client  30s
                timeout server  600s

                # HAProxy will retry only on conn-failure or empty-response conditions
                retry-on conn-failure empty-response
                retries 2


tspoe.conf:
[tspoe]
spoe-agent tspoe-agent
    groups tspoeai-reqcap-grp
    groups tspoeai-rescap-grp
    option var-prefix tspoe
    timeout hello 2s
    timeout idle  2m
    timeout processing 5s
    use-backend tspoe
    log global

spoe-group tspoeai-reqcap-grp
    messages tspoeai-reqcap

spoe-group tspoeai-rescap-grp
    messages tspoeai-rescap

spoe-message tspoeai-reqcap
    args ip=src
    args port=src_port
    args path=path
    args base=base
    args query=query
    args ssl_fc=ssl_fc
    args method=method
    args body=req.body
    args hdrs=req.hdrs_bin

spoe-message tspoeai-rescap
    args start_timestamp=var(sess.tspoe.start_timestamp)
    args trace_context.traceparent=var(sess.tspoe.trace_context.traceparent)
    args url=var(sess.tspoe.http_url)
    args target=var(sess.tspoe.http_target)
    args status=status
```

### Output of `haproxy -vv`

```plain
ubuntu@ip-37-0-2-146:~$ haproxy -vv
HAProxy version 2.4.14-1ubuntu1 2022/02/28 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.14.html
Running on: Linux 5.15.0-1019-aws #23-Ubuntu SMP Wed Aug 17 18:33:13 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with OpenSSL version : OpenSSL 3.0.1 14 Dec 2021
Running on OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.39 2021-10-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace

ubuntu@ip-37-0-2-146:~$
```

### Additional Information

If i change the 
timeout processing 5s
to 
timeout processing 10ms

the problem doesn't happen. But then it doesn't provide sufficient time for the SPOE call to execute.

**Comment by capflam** [comment]
Could you try the last 2.4 release (2.4.18) to be sure it is not an already fixed bug ?

**Comment by mtraceable** [comment]
Hi @capflam We tried with 2.4.18 with the same result.

ubuntu@ip-37-0-2-146:\~$ haproxy -vv
HAProxy version 2.4.18-1ppa1\~jammy 2022/07/27 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.18.html
Running on: Linux 5.15.0-1019-aws #23-Ubuntu SMP Wed Aug 17 18:33:13 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS +OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
Running on OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.39 2021-10-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=

Available services : prometheus-exporter
Available filters :
	[  OT] opentracing
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace

ubuntu@ip-37-0-2-146:\~$ haproxy -v
HAProxy version 2.4.18-1ppa1\~jammy 2022/07/27 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.18.html
Running on: Linux 5.15.0-1019-aws #23-Ubuntu SMP Wed Aug 17 18:33:13 UTC 2022 x86_64

ubuntu@ip-37-0-2-146:~$ ps -eaf|grep  -i hap
root       19871       1  0 16:52 ?        00:00:00 /usr/sbin/haproxy -sf 19873 19874 19875 19876 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy    19873   19871  0 16:52 ?        00:00:04 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy    19874   19871  0 16:52 ?        00:00:07 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy    19875   19871  1 16:52 ?        00:00:19 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy    19876   19871  2 16:52 ?        00:00:29 /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy    19897   19871  0 16:56 ?        00:00:02 /usr/sbin/haproxy -sf 19873 19874 19875 19876 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy    19898   19871 10 16:56 ?        00:02:03 /usr/sbin/haproxy -sf 19873 19874 19875 19876 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy    19899   19871  2 16:56 ?        00:00:26 /usr/sbin/haproxy -sf 19873 19874 19875 19876 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
haproxy    19900   19871  0 16:56 ?        00:00:08 /usr/sbin/haproxy -sf 19873 19874 19875 19876 -x /run/haproxy/admin.sock -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock

**Comment by mtraceable** [comment]
Cut down version of reproducible configurations using the SPOE library at https://github.com/negasus/haproxy-spoe-go attached.
[Archive.zip](https://github.com/haproxy/haproxy/files/9619754/Archive.zip)

**Comment by capflam** [comment]
Thanks for testing. I'll investigate. However, in the mean time, it could be usefull to get info from an old workers. Via the master socket :
```
echo 'show proc' | socat -  /run/haproxy-master.sock
# Get <pid> of an old worker
echo '@!<pid> show sess all' | socat -  /run/haproxy-master.sock
echo '@!<pid> show fd' | socat -  /run/haproxy-master.sock
```

**Comment by mtraceable** [comment]
We were also able to reproduce this on 2.0.29. Here is the output from the 2.0.29 environment:

HA-Proxy version 2.0.29-1ppa1\~focal 2022/05/14 - https://haproxy.org/
Ubuntu 20.04 focal
ubuntu@ip-37-0-2-35:\~$ echo 'show proc' | sudo socat -  /run/haproxy-master.sock
#<PID>          <type>          <relative PID>  <reloads>       <uptime>        <version>      
17528           master          0               2               0d00h30m24s     2.0.29-1ppa1\~focal
# workers
17646           worker          1               0               0d00h04m06s     2.0.29-1ppa1\~focal
17647           worker          2               0               0d00h04m06s     2.0.29-1ppa1\~focal
17648           worker          3               0               0d00h04m06s     2.0.29-1ppa1\~focal
17649           worker          4               0               0d00h04m06s     2.0.29-1ppa1\~focal
# old workers
17570           worker          [was: 1]        1               0d00h21m15s     2.0.29-1ppa1\~focal
17531           worker          [was: 2]        2               0d00h30m24s     2.0.29-1ppa1\~focal
# programs

ubuntu@ip-37-0-2-35:\~$ echo '@!17531 show sess all' | sudo socat -  /run/haproxy-master.sock
0x55762f7361f0: [22/Sep/2022:14:42:16.359579] id=118496 proto=tcpv4 source=37.0.2.152:54024
  flags=0x10080e, conn_retries=3, srv_conn=(nil), pend_pos=(nil) waiting=0
  frontend=localhost (id=3 mode=http), listener=? (id=1) addr=37.0.2.35:80
  backend=nodes (id=4 mode=http) addr=37.0.2.35:54202
  server=javaechoserver2 (id=2) addr=37.0.2.55:8080
  task=0x55762f5e6ed0 (state=0x00 nice=0 calls=8 rate=0 exp=<NEVER> tmask=0x1 age=22m46s)
  txn=0x55762f7368b0 flags=0x80300000 meth=3 status=200 req.st=MSG_DONE rsp.st=MSG_BODY
      req.f=0x0d blen=0 chnk=0 next=0
      rsp.f=0x0e blen=0 chnk=0 next=0
  si[0]=0x55762f736498 (state=EST flags=0x280008 endp0=CS:0x55762f528040 exp=<NEVER> et=0x000 sub=0)
  si[1]=0x55762f7364f0 (state=CLO flags=0x2280110 endp1=CS:0x55762f53aa50 exp=<NEVER> et=0x000 sub=0)
  co0=0x55762f4904b0 ctrl=tcpv4 xprt=RAW mux=H1 data=STRM target=LISTENER:0x55762f360b10
      flags=0x00243300 fd=363 fd.state=22 fd.cache=0 updt=0 fd.tmask=0x1
      cs=0x55762f528040 csf=0x00105000 ctx=0x55762f436410
  co1=0x55762f3ec8a0 ctrl=tcpv4 xprt=RAW mux=H1 data=STRM target=SERVER:0x55762f368d70
      flags=0x002c3300 fd=48 fd.state=22 fd.cache=0 updt=0 fd.tmask=0x1
      cs=0x55762f53aa50 csf=0x000050a0 ctx=0x55762f5874a0
  req=0x55762f736200 (f=0x68842020 an=0x48000 pipe=0 tofwd=0 total=5487)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x55762f736208 data=(nil) o=0 p=0 req.next=0 i=0 size=0
      htx=0x55762ee92fc0 flags=0x0 size=0 data=0 used=0 wrap=NO extra=0
  res=0x55762f736260 (f=0xe0000020 an=0x33000000 pipe=0 tofwd=0 total=5444)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x55762f736268 data=0x7f666f1e2010 o=0 p=0 rsp.next=0 i=131072 size=131072
      htx=0x7f666f1e2010 flags=0x0 size=131024 data=5444 used=13 wrap=NO extra=0
0x55762f54ed20: [22/Sep/2022:14:42:18.083527] id=119974 proto=tcpv4 source=37.0.2.19:33358
  flags=0x10080e, conn_retries=3, srv_conn=(nil), pend_pos=(nil) waiting=0
  frontend=localhost (id=3 mode=http), listener=? (id=1) addr=37.0.2.35:80
  backend=nodes (id=4 mode=http) addr=37.0.2.35:59260
  server=javaechoserver1 (id=1) addr=37.0.2.173:8080
  task=0x55762f4229e0 (state=0x00 nice=0 calls=8 rate=0 exp=<NEVER> tmask=0x1 age=22m44s)
  txn=0x55762f54f3e0 flags=0x80300000 meth=3 status=200 req.st=MSG_DONE rsp.st=MSG_BODY
      req.f=0x0d blen=0 chnk=0 next=0
      rsp.f=0x0e blen=0 chnk=0 next=0
  si[0]=0x55762f54efc8 (state=EST flags=0x280008 endp0=CS:0x55762f718b90 exp=<NEVER> et=0x000 sub=0)
  si[1]=0x55762f54f020 (state=CLO flags=0x2280110 endp1=CS:0x55762f847220 exp=<NEVER> et=0x000 sub=0)
  co0=0x55762f5cf480 ctrl=tcpv4 xprt=RAW mux=H1 data=STRM target=LISTENER:0x55762f360b10
      flags=0x00243300 fd=1202 fd.state=22 fd.cache=0 updt=0 fd.tmask=0x1
      cs=0x55762f718b90 csf=0x00105000 ctx=0x55762f60c220
  co1=0x55762f4135b0 ctrl=tcpv4 xprt=RAW mux=H1 data=STRM target=SERVER:0x55762f365f30
      flags=0x002c3300 fd=731 fd.state=22 fd.cache=0 updt=0 fd.tmask=0x1
      cs=0x55762f847220 csf=0x000050a0 ctx=0x55762f4c17d0
  req=0x55762f54ed30 (f=0x68842020 an=0x48000 pipe=0 tofwd=0 total=5488)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x55762f54ed38 data=(nil) o=0 p=0 req.next=0 i=0 size=0
      htx=0x55762ee92fc0 flags=0x0 size=0 data=0 used=0 wrap=NO extra=0
  res=0x55762f54ed90 (f=0xe0000020 an=0x33000000 pipe=0 tofwd=0 total=5445)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x55762f54ed98 data=0x7f66702a4010 o=0 p=0 rsp.next=0 i=131072 size=131072
      htx=0x7f66702a4010 flags=0x0 size=131024 data=5445 used=13 wrap=NO extra=0
0x55762f8d3360: [22/Sep/2022:15:05:02.250642] id=120873 proto=sockpair source=unix:3
  flags=0x8, conn_retries=0, srv_conn=(nil), pend_pos=(nil) waiting=0
  frontend=GLOBAL (id=0 mode=tcp), listener=? (id=3) addr=37.0.2.35:80
  backend=<NONE> (id=-1 mode=-)
  server=<NONE> (id=-1)
  task=0x55762f406e50 (state=0x00 nice=-64 calls=1 rate=0 exp=<NEVER> tmask=0x1 age=0s)
  si[0]=0x55762f8d3608 (state=EST flags=0x280008 endp0=CS:0x55762f60b100 exp=<NEVER> et=0x000 sub=0)
  si[1]=0x55762f8d3660 (state=EST flags=0x204018 endp1=APPCTX:0x55762f8a8880 exp=<NEVER> et=0x000 sub=0)
  co0=0x55762f6e1530 ctrl=sockpair xprt=RAW mux=PASS data=STRM target=LISTENER:0x55762f357b40
      flags=0x00241300 fd=11 fd.state=22 fd.cache=0 updt=0 fd.tmask=0x1
      cs=0x55762f60b100 csf=0x00009000 ctx=0x55762f60b120
  app1=0x55762f8a8880 st0=7 st1=0 st2=3 applet=<CLI> tmask=0x1 nice=-64 calls=1 rate=0 cpu=0 lat=0
  req=0x55762f8d3370 (f=0xc4c220 an=0x0 pipe=0 tofwd=0 total=14)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x55762f8d3378 data=0x7f667144d010 o=0 p=0 req.next=0 i=0 size=131072
  res=0x55762f8d33d0 (f=0x80008002 an=0x0 pipe=0 tofwd=-1 total=3572)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x55762f8d33d8 data=0x7f6671f21010 o=3572 p=3572 rsp.next=0 i=0 size=131072

ubuntu@ip-37-0-2-35:\~$ echo '@!17531 show fd' | sudo socat -  /run/haproxy-master.sock
      6 : st=0x05(R:PrA W:pra) ev=0x01(heopI) [lc] cnext=-15 cprev=-2 tmask=0x1 umask=0x0 owner=0x55762f357b40 iocb=0x55762ecfda40(mworker_accept_wrapper)
      7 : st=0x05(R:PrA W:pra) ev=0x00(heopi) [lc] cnext=-3 cprev=0 tmask=0x1 umask=0x0 owner=0x55762ecd63a0 iocb=0x55762ecd63a0(poller_pipe_io_handler)
     11 : st=0x22(R:pRa W:pRa) ev=0x01(heopI) [lc] cnext=-3 cprev=-2 tmask=0x1 umask=0x0 owner=0x55762f8a8880 iocb=0x55762ecc5020(conn_fd_handler) back=0 cflg=0x00241300 fe=GLOBAL mux=PASS ctx=0x55762f4a3c40
     48 : st=0x22(R:pRa W:pRa) ev=0x01(heopI) [lc] cnext=-3 cprev=-2 tmask=0x1 umask=0x0 owner=0x55762f3ec8a0 iocb=0x55762ecc5020(conn_fd_handler) back=1 cflg=0x002c3300 sv=nodes/javaechoserver2 mux=H1 ctx=0x55762f3ecaf0 h1c.flg=0x4100 .sub=0 .ibuf=0@(nil)+0/0 .obuf=0@(nil)+0/0 h1s=0x55762f5874a0 h1s.flg=0x48c8 .req.state=MSG_DONE .res.state=MSG_DONE .meth=POST status=200 .cs.flg=0x000050a0 .cs.data=0x55762f7364f0
    363 : st=0x22(R:pRa W:pRa) ev=0x11(HeopI) [lc] cnext=-3 cprev=-2 tmask=0x1 umask=0x0 owner=0x55762f4904b0 iocb=0x55762ecc5020(conn_fd_handler) back=0 cflg=0x00243300 fe=localhost mux=H1 ctx=0x55762f71bbc0 h1c.flg=0x40 .sub=0 .ibuf=0@(nil)+0/0 .obuf=0@(nil)+0/0 h1s=0x55762f436410 h1s.flg=0x98 .req.state=MSG_DONE .res.state=MSG_RPBEFORE .meth=POST status=0 .cs.flg=0x00105000 .cs.data=0x55762f736498
    731 : st=0x22(R:pRa W:pRa) ev=0x01(heopI) [lc] cnext=-3 cprev=-2 tmask=0x1 umask=0x0 owner=0x55762f4135b0 iocb=0x55762ecc5020(conn_fd_handler) back=1 cflg=0x002c3300 sv=nodes/javaechoserver1 mux=H1 ctx=0x55762f764930 h1c.flg=0x4100 .sub=0 .ibuf=0@(nil)+0/0 .obuf=0@(nil)+0/0 h1s=0x55762f4c17d0 h1s.flg=0x48c8 .req.state=MSG_DONE .res.state=MSG_DONE .meth=POST status=200 .cs.flg=0x000050a0 .cs.data=0x55762f54f020
   1202 : st=0x22(R:pRa W:pRa) ev=0x11(HeopI) [lc] cnext=-3 cprev=-2 tmask=0x1 umask=0x0 owner=0x55762f5cf480 iocb=0x55762ecc5020(conn_fd_handler) back=0 cflg=0x00243300 fe=localhost mux=H1 ctx=0x55762f5ade50 h1c.flg=0x40 .sub=0 .ibuf=0@(nil)+0/0 .obuf=0@(nil)+0/0 h1s=0x55762f60c220 h1s.flg=0x98 .req.state=MSG_DONE .res.state=MSG_RPBEFORE .meth=POST status=0 .cs.flg=0x00105000 .cs.data=0x55762f54efc8

ubuntu@ip-37-0-2-35:\~$ echo '@!17570 show sess all' | sudo socat -  /run/haproxy-master.sock
0x5584ddb170c0: [22/Sep/2022:14:59:25.621284] id=850988 proto=tcpv4 source=37.0.1.239:56570
  flags=0x10080e, conn_retries=3, srv_conn=(nil), pend_pos=(nil) waiting=0
  frontend=localhost (id=3 mode=http), listener=? (id=1) addr=37.0.2.35:80
  backend=nodes (id=4 mode=http) addr=37.0.2.35:47608
  server=javaechoserver2 (id=2) addr=37.0.2.55:8080
  task=0x5584dd824b10 (state=0x00 nice=0 calls=7 rate=0 exp=<NEVER> tmask=0x1 age=8m11s)
  txn=0x5584ddb17760 flags=0x80300000 meth=3 status=200 req.st=MSG_DONE rsp.st=MSG_BODY
      req.f=0x0d blen=0 chnk=0 next=0
      rsp.f=0x0e blen=0 chnk=0 next=0
  si[0]=0x5584ddb17368 (state=EST flags=0x280008 endp0=CS:0x5584dd992ad0 exp=<NEVER> et=0x000 sub=0)
  si[1]=0x5584ddb173c0 (state=CLO flags=0x2280110 endp1=CS:0x5584dd8b9eb0 exp=<NEVER> et=0x000 sub=0)
  co0=0x5584dd9f3450 ctrl=tcpv4 xprt=RAW mux=H1 data=STRM target=LISTENER:0x5584dd626b20
      flags=0x00243300 fd=560 fd.state=22 fd.cache=0 updt=0 fd.tmask=0x1
      cs=0x5584dd992ad0 csf=0x00105000 ctx=0x5584dd855b20
  co1=0x5584ddc363c0 ctrl=tcpv4 xprt=RAW mux=H1 data=STRM target=SERVER:0x5584dd62ed80
      flags=0x002c3300 fd=1438 fd.state=22 fd.cache=0 updt=0 fd.tmask=0x1
      cs=0x5584dd8b9eb0 csf=0x000050a0 ctx=0x5584dd7be8d0
  req=0x5584ddb170d0 (f=0x68842020 an=0x48000 pipe=0 tofwd=0 total=5488)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x5584ddb170d8 data=(nil) o=0 p=0 req.next=0 i=0 size=0
      htx=0x5584dd12bfc0 flags=0x0 size=0 data=0 used=0 wrap=NO extra=0
  res=0x5584ddb17130 (f=0xe0000020 an=0x33000000 pipe=0 tofwd=0 total=5445)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x5584ddb17138 data=0x7fe39630f010 o=0 p=0 rsp.next=0 i=131072 size=131072
      htx=0x7fe39630f010 flags=0x0 size=131024 data=5445 used=13 wrap=NO extra=0
0x5584ddae05f0: [22/Sep/2022:15:07:36.018258] id=853152 proto=sockpair source=unix:2
  flags=0x8, conn_retries=0, srv_conn=(nil), pend_pos=(nil) waiting=0
  frontend=GLOBAL (id=0 mode=tcp), listener=? (id=2) addr=37.0.2.35:80
  backend=<NONE> (id=-1 mode=-)
  server=<NONE> (id=-1)
  task=0x5584dd860310 (state=0x00 nice=-64 calls=1 rate=0 exp=<NEVER> tmask=0x1 age=0s)
  si[0]=0x5584ddae0898 (state=EST flags=0x280008 endp0=CS:0x5584dd8d75d0 exp=<NEVER> et=0x000 sub=0)
  si[1]=0x5584ddae08f0 (state=EST flags=0x204018 endp1=APPCTX:0x5584dd963820 exp=<NEVER> et=0x000 sub=0)
  co0=0x5584ddb9f780 ctrl=sockpair xprt=RAW mux=PASS data=STRM target=LISTENER:0x5584dd61d5f0
      flags=0x00241300 fd=8 fd.state=22 fd.cache=0 updt=0 fd.tmask=0x1
      cs=0x5584dd8d75d0 csf=0x00009000 ctx=0x5584dd8d75f0
  app1=0x5584dd963820 st0=7 st1=0 st2=3 applet=<CLI> tmask=0x1 nice=-64 calls=1 rate=0 cpu=0 lat=0
  req=0x5584ddae0600 (f=0xc4c220 an=0x0 pipe=0 tofwd=0 total=14)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x5584ddae0608 data=0x5584ddf80640 o=0 p=0 req.next=0 i=0 size=131072
  res=0x5584ddae0660 (f=0x80008002 an=0x0 pipe=0 tofwd=-1 total=1786)
      an_exp=<NEVER> rex=<NEVER> wex=<NEVER>
      buf=0x5584ddae0668 data=0x7fe397266010 o=1786 p=1786 rsp.next=0 i=0 size=131072

ubuntu@ip-37-0-2-35:\~$ echo '@!17570 show fd' | sudo socat -  /run/haproxy-master.sock
      5 : st=0x05(R:PrA W:pra) ev=0x00(heopi) [lc] cnext=-3 cprev=0 tmask=0x1 umask=0x0 owner=0x5584dcf6f3a0 iocb=0x5584dcf6f3a0(poller_pipe_io_handler)
      6 : st=0x05(R:PrA W:pra) ev=0x01(heopI) [lc] cnext=-12 cprev=-2 tmask=0x1 umask=0x0 owner=0x5584dd61d5f0 iocb=0x5584dcf96a40(mworker_accept_wrapper)
      8 : st=0x22(R:pRa W:pRa) ev=0x01(heopI) [lc] cnext=-3 cprev=-2 tmask=0x1 umask=0x0 owner=0x5584dd963820 iocb=0x5584dcf5e020(conn_fd_handler) back=0 cflg=0x00241300 fe=GLOBAL mux=PASS ctx=0x5584dd6be9f0
    560 : st=0x22(R:pRa W:pRa) ev=0x11(HeopI) [lc] cnext=-3 cprev=-2 tmask=0x1 umask=0x0 owner=0x5584dd9f3450 iocb=0x5584dcf5e020(conn_fd_handler) back=0 cflg=0x00243300 fe=localhost mux=H1 ctx=0x5584dd9647b0 h1c.flg=0x40 .sub=0 .ibuf=0@(nil)+0/0 .obuf=0@(nil)+0/0 h1s=0x5584dd855b20 h1s.flg=0x98 .req.state=MSG_DONE .res.state=MSG_RPBEFORE .meth=POST status=0 .cs.flg=0x00105000 .cs.data=0x5584ddb17368
   1438 : st=0x22(R:pRa W:pRa) ev=0x11(HeopI) [lc] cnext=-3 cprev=-2 tmask=0x1 umask=0x0 owner=0x5584ddc363c0 iocb=0x5584dcf5e020(conn_fd_handler) back=1 cflg=0x002c3300 sv=nodes/javaechoserver2 mux=H1 ctx=0x5584dda57570 h1c.flg=0x4100 .sub=0 .ibuf=0@(nil)+0/0 .obuf=0@(nil)+0/0 h1s=0x5584dd7be8d0 h1s.flg=0x48c8 .req.state=MSG_DONE .res.state=MSG_DONE .meth=POST status=200 .cs.flg=0x000050a0 .cs.data=0x5584ddb173c0

**Comment by mtraceable** [comment]
Also we were not able to reproduce this problem on 2.0.13-2ubuntu0.5.

**Comment by capflam** [comment]
Thanks for these info, I'll take a look.

**Comment by capflam** [comment]
So, indeed, some sessions are blocked on the response analysis. I don't know for now how it really happens. But there is an issue with the SPOE on stop/reload. The session is not notified the processing was aborted. I must dig in that direction.

**Comment by mtraceable** [comment]
Thanks @capflam. Please let me know if I can provide any other information regarding this.

**Comment by mtraceable** [comment]
Hi @capflam can I help provide any other information? This is affecting our production deployment and we were wondering if there is any fix or workaround that you can think of for the issue?

**Comment by capflam** [comment]
Sorry, I'm still unable to understand how it is possible. I read the code to try to find how the processing expiration date may be ignored or not properly set. But for now, it is a mystery for me. However there is something odd with your configuration. You are defining the connect/client/server/http-keep-alive timeouts in the defaults section at the end of your configuration file. It means no timeout are defined on frontends and backends above.

In your traces, the worker `17531` is stopping since 30min while the stalled sessions are a little less than 23min old. So without the SPOE, an inactive client may block infinitely your workers on soft-stop. 

So I'm still digging. At the end, if I'm still unable to find anything strange, I would ask you to kill an old worker with a SIGABRT to generate a core file and be able to do a postmortem analysis with GDB. But it is a painful process and I need a bit more time to read the code. 

As a workaround, you may set a `hard-stop-after` timeout. It is not satisfying but may help for now.

**Comment by tim-mwangi** [comment]
Hi @capflam , I’m working with @mtraceable on this issue. Quick update about the suggestion to put timeouts in the frontends and backends. Specifying either the backend config `timeout server <duration>` or global config `hard-stop-after <duration>` fixes the issue and the workers exit after `<duration>`. So for backend config timeout server 20s, the workers exit after 20s. Similarly for `hard-stop-after <duration>` global config, the workers exit after `<duration>`. We are now specifying explicit timeouts for the frontends and backends. client for frontends and connect and server for backends.

**Comment by capflam** [comment]
Thanks for the update. Indeed, timeouts are required to avoid troubles with idle clients. I'm glad it did the trick for your use-case. However, it seems there is something odd with the SPOE. As I said, I don't know if it is related to the soft-stop or not. But the response analysis is blocked and the processing timeout on the SPOE side seems to be disabled for an unknown reason.

I  didn't have the time to work on it till now. So I'll keep the issue open as a reminder. I must definitely work on it.

**Comment by mtraceable** [comment]
Hi @capflam! Did you get a chance to dig this further? Please let us know if I can help in anyway.

**Comment by capflam** [comment]
No not really. I must carefully read the code to figure out how it is possible. It is a painful and time consuming task. For now, nothing obvious hit me. However, you may try the 2.4.19. It contains a SPOE fix. At first glance, it is unrelated but who knows... But I'm curious. I understood that setting the timeouts solved your problem. Did you hit it again, even with timeouts ?

**Comment by mtraceable** [comment]
Hi @capflam we resolved the issue of stuck processes with the timeout. We are still seeing an issue where in the reload case some of the response information is not coming to us through the SPOE so we are missing some data. And we see the following in the logs:

`TRACE: 2022/10/06 18:04:19 log.go:31: error: ack frame write failed: cannot write frame to connection: write tcp 127.0.0.1:5444->127.0.0.1:59654: use of closed network connection
TRACE: 2022/10/06 18:04:19 log.go:31: error: ack frame write failed: cannot write frame to connection: write tcp 127.0.0.1:5444->127.0.0.1:38392: use of closed network connection
TRACE: 2022/10/06 18:04:19 log.go:31: error: ack frame write failed: cannot write frame to connection: write tcp 127.0.0.1:5444->127.0.0.1:38402: use of closed network connection`

**Comment by capflam** [comment]
These messages suggest that HAProxy has closed the SPOE connections on reloads. It may be a side effect of the "hard-stop-after" option. You may check if the errors happens after the hard-stop-after delay on reload. But in this case, it is a bit strange because there was no SPOE connection in traces you provided. You may try to enable the SPOE logs and also try with the 2.4.19.

**Comment by pasztorl** [comment]
I've just tested this with 2.4.22 and the problem still exists. I'm not using SPOE in my config, so SPOE may not revelant(?).

**Comment by Darlelet** [comment]
@pasztorl Interesting, are you able to reproduce it consistently? Do you think you can share some details about your setup / configuration leading to this state? Thanks
@wlallemand it reminded me of the case you described this morning

**Comment by TreZc0** [comment]
We face the same issue on the latest Debian 11 native build, 2.2.9.

We do not use SPOE in our config, but a fairly standard config with frontend and backends, ACLs, and header adjustments. 

On service reload, the old process is not killed / exited properly, and we have two process running with different configs.

**Comment by pasztorl** [comment]
@Darlelet we discovered that it was a websocket/timeout problem, so not related this issue.

**Comment by wtarreau** [comment]
@pasztorl OK thanks for letting us know.

**Comment by TimWolla** [comment]
> We face the same issue on the latest Debian 11 native build, 2.2.9.

@TreZc0 HAProxy 2.2.9 is extremely old. On Debian I recommend you use Vincent Bernat's packages from https://haproxy.debian.net/.

**Comment by mtraceable** [comment]
Wanted to update on this. This continues to be an issue even on 2.4.17.

**Comment by capflam** [comment]
The 2.4.17 is pretty old. You should try the 2.4.23. But to be sure, it is the stalled sessions issue or the  connections closure issue because of the hard-stop-after option ? It is related because the hard-stop-after option will kill the stalled sessions. But the issue is a bit old for me now, I need some help to put the pieces together :)

**Comment by tim-mwangi** [comment]
Hey @capflam, getting back to this one. We have not tested v2.4.23 yet but will do so soon and report if we can still reproduce it. I have some questions concerning another issue that happens when doing response capture and sending response data to an SPOE agent. The version concerned is v2.4.17

- We have observed that sometimes we hit the file descriptor(FD) soft limit given by `ulimit -n` and get a `Too many open files` error message when trying to create a new connection between the Haproxy and the SPOE agent. This happens when have response capture enabled eg. `http-response send-spoe-group tspoe tspoeai-rescap-grp` and several `systemctl reload haproxy` commands are issued. Do you have any idea if this would be related to the hanging processes issue?
- Given that the architecture has changed from worker processes to threads, how likely is that the issue persists if it is indeed related to the hanging processes?

Our team has not been able to reproduce the FD limit issue internally but it is happening in an environment running our SPOE agent but we do not have access to it.

Thanks!

**Comment by capflam** [comment]
Sorry its seems I missed your reply. If you hit the FD limit it means there is too many connections opened on your agent. An issue was fixed in 2.4.23 to not create too many connections on sporadic burst (d1e1aa01 "BUG/MEDIUM: spoe: Don't start new applet if there are enough idle ones"). So it may be related. However, it should be unrelated to reloads except if the new process tries suddenly to open many new connections and exceeds the FD limit. You can try be setting a `maxconnrate` in your SPOE configuration and/or a `maxconn` in your SPOE bakend to limit the number of connections.

**Comment by capflam** [comment]
It is just a ping to know if you are still affected by this issue with recent versions ?

**Comment by alebsys** [comment]
> It is just a ping to know if you are still affected by this issue with recent versions ?

We have an identical problem on:
- hapee version: 2.8.0-1.0.0-311.452
- Linux  5.10.0-20-amd64 SMP Debian 5.10.158-2 (2022-12-13) x86_64 GNU/Linux

**Comment by capflam** [comment]
Thanks @alebsys. I recently pushed 2 fixes about the SPOE that may help:

* 6547b14292: BUG/MINOR: spoe: Be sure to be able to quickly close IDLE applets on soft-stop
* 3c066b1e34: BUG/MEDIUM: spoe: Don't rely on stream's expiration to detect processing timeout

There were not backported yet. But this will be done soon.

**Comment by capflam** [comment]
FYI, commits above were backported 2.8.

**Comment by capflam** [comment]
I don't know if the issue is still relevant or not. But it is quite old now. I'm closing. Feel free to reopen it if there is still a bug. Thanks !
